### PR TITLE
Feature/Added prop for passing function when close button is clicked

### LINKED ISSUE #1117: Feature: prop for passing function when close button is clicked
**Do you want to request a *feature* or report a *bug*?**
Feature
**What is the current behavior?**
The current behaviour is that I have to create a new button so I can pass an onClick listener, but the one by default its fine.  

Would be to easy to add a new prop like: 
onButtonClose | function | Called when the notification close button is clicked.
-- | -- | --

### PR DESCRIPTION
**Pull request by sbsangu**
### Fixes

- Fixes #1117 

### Summary

Added an `onButtonClose` prop to the `Toast` component to allow users to define a callback function that gets called when the close button of a toast is clicked. This enhancement improves the flexibility and usability of the toast component by allowing custom behavior to be executed when the close button is pressed.

### Changes

- Updated the `Toast` component to accept and handle a new `onButtonClose` prop.
- Updated the `ToastContainer` component to pass the `onButtonClose` prop to each `Toast` component.
- Updated TypeScript definitions to include the new `onButtonClose` prop.

### Checklist

- [x] Forked the repository and created a new branch from `master`.
- [x] Ran `yarn` in the repository root.
- [x] Added tests to cover the new functionality.
- [x] Ensured the test suite passes (`yarn test`).
- [x] Ran `yarn start` to test changes in the playground.
- [x] Updated the readme if needed.
- [x] Updated the TypeScript definitions if needed.
- [x] Formatted the code with [prettier](https://github.com/prettier/prettier) (`yarn prettier-all`).
- [x] Ensured the code lints (`yarn lint:fix`).

### Additional Information

This feature allows users to pass a callback function that will be called when the toast's close button is clicked, enabling more custom and dynamic behaviors when a toast is dismissed.

### CODE REVIEW

**Review by CalvinJamesHeath — CHANGES_REQUESTED** [review]
Excellent!

**Comment by sbsangu** [comment]
Could you please let me know if this pull request resolves the issue? 
Additionally, do you think it is ready to be merged?

File: README.md [hunk 1]
```diff
@@ -107,6 +107,12 @@ Support this project with your organization. Your logo will show up here with a
 <a href="https://opencollective.com/react-toastify/organization/8/website"><img src="https://opencollective.com/react-toastify/organization/8/avatar.svg"></a>
 <a href="https://opencollective.com/react-toastify/organization/9/website"><img src="https://opencollective.com/react-toastify/organization/9/avatar.svg"></a>
 
+### ToastContainer Props
```
Comment (zizhengtai) [inline README.md:110]: I don't think this should be added here, the README seems to contain only high-level information about the project and not detailed release notes.

File: package-lock.json
Comment (zizhengtai) [inline package-lock.json:1]: Can we delete this? The project is not using package-lock.json.

File: yarn.lock
Comment (zizhengtai) [inline yarn.lock:1]: Can you revert these changes? There doesn't seem to be any good reason for it.

**Review by zizhengtai — CHANGES_REQUESTED** [review]

**Pull request closed without merging**